cpp/ql/src/semmle/code/cpp/security/FlowSources.qll

```diff
@@ -42,3 +52,49 @@ private class TaintedParameterSource extends RemoteFlowSource {
 
   override string getSourceType() { result = sourceType }
 }
+
+private class LocalReturnSource extends LocalFlowSource {
+  string sourceType;
+
+  LocalReturnSource() {
+    exists(LocalFlowFunction func, CallInstruction instr, FunctionOutput output |
+      asInstruction() = instr and
+      instr.getStaticCallTarget() = func and
+      func.hasLocalFlowSource(output, sourceType) and
+      (
+        output.isReturnValue()
+        or
+        output.isReturnValueDeref()
+      )
+    )
+  }
+
+  override string getSourceType() { result = sourceType }
+}
+
+private class LocalParameterSource extends LocalFlowSource {
+  string sourceType;
+
+  LocalParameterSource() {
+    exists(LocalFlowFunction func, WriteSideEffectInstruction instr, FunctionOutput output |
+      asInstruction() = instr and
+      instr.getPrimaryInstruction().(CallInstruction).getStaticCallTarget() = func and
+      func.hasLocalFlowSource(output, sourceType) and
+      output.isParameterDeref(instr.getIndex())
```

Comment:
This modelling doesn't handle `output.isQualifierObject()`, and I don't see that handled anywhere else. Does `instr.getIndex() = -1` hold for the qualifier side effect? If so, perhaps we should add a helper predicate so we can write `output.isParameterDerefOrQualifierObject(index)` or something like that.

Reply:
Indeed, `instr.getIndex() = -1` holds for the qualifier side effect, and so we should special-case `output.isQualifierObject()` when the index is `-1`. I suppose this is related to https://github.com/github/codeql-c-analysis-team/issues/186.

Reply:
4cb25d8e18bfd283c67474486498a9d1e39e0803 adds the predicate and uses it in the charpreds of `RemoteParameterSource` and `LocalParameterSource`.